### PR TITLE
feat: allow auditors to read template insights

### LINKED ISSUE #10627: allow auditors to query `insights/user-latency`
Getting `{"message":"Resource not found or you do not have access to this resource"}` when querying `insights/user-latency` with an Auditor role. Could we consider allow Auditor role for this endpoint please

### PR DESCRIPTION
**Pull request by johnstcn**
Fixes https://github.com/coder/coder/issues/10627

- Adds a template_insights pseudo-resource
- Grants auditor and template admin roles read access on template_insights
- Updates existing RBAC checks to check for read template_insights, falling back to template update permissions
- Updates TemplateLayout to show Insights tab if can read template_insights or can update template

### CODE REVIEW

File: coderd/rbac/roles.go [hunk 1]
```diff
@@ -195,6 +196,8 @@ func ReloadBuiltinRoles(opts *RoleOptions) {
 			ResourceGroup.Type:        {ActionRead},
 			// Org roles are not really used yet, so grant the perm at the site level.
 			ResourceOrganizationMember.Type: {ActionRead},
+			// Template admins can read all template insights data
```
Comment (mtojek) [inline coderd/rbac/roles.go:199]: nit: template admins and auditors?

File: coderd/database/dbauthz/dbauthz.go [hunk 2]
```diff
@@ -1344,64 +1348,72 @@ func (q *querier) GetTemplateDAUs(ctx context.Context, arg database.GetTemplateD
...
+		if len(arg.TemplateIDs) == 0 {
+			if err := q.authorizeContext(ctx, rbac.ActionUpdate, rbac.ResourceTemplate.All()); err != nil {
+				return nil, err
+			}
 		}
 	}
 	return q.db.GetTemplateInsightsByInterval(ctx, arg)
 }
 
 func (q *querier) GetTemplateInsightsByTemplate(ctx context.Context, arg database.GetTemplateInsightsByTemplateParams) ([]database.GetTemplateInsightsByTemplateRow, error) {
```
Comment (mtojek) [inline coderd/database/dbauthz/dbauthz.go:1392]: Some of these methods are used only by metrics collector, so in theory you don't need to assign auditor permissions.

File: site/src/pages/TemplatePage/TemplateLayout.tsx [hunk 3]
```diff
@@ -68,7 +74,8 @@ export const TemplateLayout: FC<{ children?: JSX.Element }> = ({
     queryKey: ["template", templateName],
     queryFn: () => fetchTemplate(orgId, templateName),
   });
-  const shouldShowInsights = data?.permissions?.canUpdateTemplate;
+  const shouldShowInsights =
+    data?.permissions?.canUpdateTemplate || data?.permissions?.canReadInsights;
```
Comment (mtojek) [inline site/src/pages/TemplatePage/TemplateLayout.tsx:78]: maybe drop a comment here (why 2 permissions instead of 1) as somebody may find it redundant.

**Review by mtojek — APPROVED** [review]
Mostly nit picks 👍

File: coderd/database/dbauthz/dbauthz.go (hunk 2 above)
Comment (johnstcn) [reply coderd/database/dbauthz/dbauthz.go:1392]: I'll take another pass and remove the unnecessary perms.

File: coderd/rbac/roles.go (hunk 1 above)
Comment (johnstcn) [reply coderd/rbac/roles.go:199]: The context here is for the template admin role, so I think the comment is still valid.

Comment (mtojek) [reply coderd/rbac/roles.go:199]: Sorry, didn't notice that. Thanks for clarifying.

File: coderd/database/dbauthz/dbauthz.go (hunk 2 above)
Comment (johnstcn) [reply coderd/database/dbauthz/dbauthz.go:1397]: I ended up switching `GetTemplateAppInsightsByTemplate` and `GetTemplateInsightsByTemplate` to only check `rbac.ResourceTemplateInsights` as they're only used by metrics collector and it's likely to be lower overhead than `rbac.ResourceTemplate.All()`. Could also have made check `rbac.ResourceSystem`, but I think this is more correct.

**Pull request merged**